kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncProcessor.java

```diff
@@ -299,12 +300,23 @@ private void processNewAsyncEvent(final AsyncEvent event) {
           "process called when already hit exception: " + fatalException);
     }
 
-    pendingEvents.add(event);
+    try {
+      pendingEvents.add(event);
+
+      maybeBackOffEnqueuingNewEventWithKey(event.inputRecordKey());
+      schedulingQueue.offer(event);
 
-    maybeBackOffEnqueuingNewEventWithKey(event.inputRecordKey());
-    schedulingQueue.offer(event);
+      executeAvailableEvents();
 
-    executeAvailableEvents();
+    } catch (final StreamsException streamsException) {
+      fatalException = streamsException.taskId().isPresent()
+          ? streamsException
+          : new StreamsException(streamsException, taskId);
+      throw fatalException;
+    } catch (final Throwable t) {
+      fatalException = new StreamsException(t, taskId);
+      throw fatalException;
```

Comment:
This block is repeated a few times -- can put it into a single method if you want to avoid duplicate code but in theory we should never touch this stuff again. 

Basically I extracted all setting and checking of the top-level `fatalException` to the top level of the three "entry points" to the AsyncProcessor class: the `#process` methods, the `#punctuate`, and the `#flushAndAwaitPendingEvents`